### DISCUSSION TITLE
Support `define:vars` and `hoisted` imports

### DISCUSSION THREAD

**Discussion by vovsemenv** [withastro/roadmap#295] [Proposal]
## The problem
Hello, astronauts 👩🏼‍🚀  Currently when you need to use server side values inside your client code you can use 
```astro
<script define:vars={{value}}>
    console.log(value);
</script>
```
It's ok for very simple cases, but if you try to import another module inside this script it won't work because `define:vars` forces `is:inline` for this tag and imported scripts not processed by astro so you use this kinda workarounds.
```astro
---
const someValue = "value";
---
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <meta name="viewport" content="width=device-width" />
  </head>
  <body>
    <!-- In astro you can't use define:vars and import inside one script so we use this workaround-->
    <script define:vars={{ someValue }}>
      window.someValue = someValue;
    </script>
    <script is:inline>
      import { doSomethingImportant } from "../utils";
      doSomethingImportant((window as any).someValue);
    </script>
  </body>
</html>
```

I think it's very common use-case for `<scripts define:vars>` and it should support hoisted imports inside

**Comment by vovsemenv** [comment]
Also @delucis mentioned that my workaround potentially buggy.
I think we really need some bulletproof solution to prevent this types of bugs
https://discord.com/channels/830184174198718474/872579324446928896/1018432947356975144
<img width="1091" alt="image" src="https://user-images.githubusercontent.com/20096510/189522256-76f8bea6-9285-443a-a2db-499d452f2d61.png">

**Comment by matthewp** [comment]
The issue is that hoisted scripts get built and bundled. So we would need to do something like this:

```astro
<script define:vars={{someValue}}>
    console.log(someValue);
</script>
```

Becomes:

```html
<script>
  (window._astroDefineVars || window._astroDefineVars = {}).someValue = "value";
</script>
```

__hoisted.bundle.js__

```js
var someValue = window._astroDefineVars.someValue;
console.log(someValue);
```

**Comment by crutchcorn** [comment]
Could a potential API be something like:


```
<script define:vars={{someValue}} global:key={uniqueKey}>
    console.log(someValue);
</script>
```

And allow `uniqueKey` to distinguish which script belongs to what page?

**Comment by ayo-run** [comment]
An option would be to extend the `define:vars` implementation and prepend the unique Astro component class as key/id to an ID passed for the component script (to make sure it is really unique across components) to be used into a common `<script type="application/json">` in the head of the document for faster querying.

So it would look like:

```html
<script type="application/json">
{
  astroJfkdlsfj: {
    someValue: 'value'
  },
  astroKfjkdsd: {
    someValue: 'anotherValue'
 }
}
</script>
```

Where `astroJfkdlsfj` and `astroKfjkdsd` are component class names + script ID in those components.

To use in the client script, we could have a function that just gets those data from the shared JSON script or even assign to variables already with same names as they have in the server script.

We could preserve more complex types as well if we use something like [`devalue`](https://github.com/Rich-Harris/devalue) but we probably want users to opt-in on that dependency

**Comment by lilnasy** [comment]
Here's the idea I'm going to be exploring over the weekend.

```astro
---
interface Props {
    x : string
}

const x = Astro.props.x
---
<script define:args={{ x }}>
import { variable } from "library"

console.log("Top level statements execute on the initial page load.")

export default ({ x }: { x : string }) => {
    console.log(`This block reruns once for every use of the component. In this run, x is ${x}`)
}

document.addEventListener('astro:page-load', () => {
    console.log("This block runs on every view transition naviagtion.")
})
</script>
```

**Reply by lilnasy** [reply]
The experimentation is promising! Full proposal here: https://github.com/withastro/roadmap/discussions/705.
